### ISSUE TITLE
Mono.Cecil.AssemblyDefinition.ReadAssembly huge RAM usage

### ISSUE THREAD

**Issue by nefcanto** [jbevain/cecil#576]
I've created a coding quality tool that uses reflection for some purposes. Given a codebase, I load assembly files in build folders and process them. There are like 30 thousand DLL files.

I had no problems with earlier versions of Mono.Cecil for this, but now, it immediately consumes like 8 Giga bytes of RAM.

I use Mono.Cecil.AssemblyDefinition.ReadAssembly(pathToAssembly) to load them.

What can be wrong?

**Comment by SimonCropp** [comment]
both `AssemblyDefinition` and `ModuleDefinition` implement `IDisposable`. are you disposing of them when you are finished with each assembly?

**Comment by jbevain** [comment]
Starting with 0.10 you do indeed need to dispose of the assemblies.

**Comment by nefcanto** [comment]
Test this code please. Get a list of assembly paths, and run this method on them both sequentially, and in parallel. The parallel mode consumes huge amount of RAM:

	private static List<Tuple<AssemblyDefinition, string>> ConvertToReflectiveAssemblies(List<string> assemblyFiles)
	{
		var temp = new List<Tuple<AssemblyDefinition, string>>();

		//for (int index = 0; index < assemblyFiles.Count; index++)
		Parallel.For(0, assemblyFiles.Count, StatisticsConfig.ParallelOptions, index =>
		{
			var assemblyFile = assemblyFiles[index];
			try
			{
				temp.Add(new Tuple<AssemblyDefinition, string>(AssemblyDefinition.ReadAssembly(assemblyFile), assemblyFile));
			}
			catch (Exception ex)
			{
				Logger.Log(ex);
				Logger.LogError($"Error reading AssemblyDefinition from {assemblyFile}");
			}
			Logger.Count(index);
			//}
		});
		var result = temp.Where(i => i.IsNotNull()).ToList();
		return result;
	}

Using Parallel to read assembly definitions:
![image](https://user-images.githubusercontent.com/12782347/62820628-46d1bc80-bb7c-11e9-889a-e055e6b197b6.png)

Reading assembly definitions sequentially:
![image](https://user-images.githubusercontent.com/12782347/62820641-7b457880-bb7c-11e9-802c-63c28ca22d01.png)



What can cause this error?

**Comment by jbevain** [comment]
I'm sorry but there's nothing actionable here for me.

The code that I'm supposed to «test» is a piece of code that reads `n` assemblies and keep them in a List in memory. If you have many assemblies, it will consume memory.

**Comment by nefcanto** [comment]
Thank you for answering. As you can see the problem is in parallelism.

I know reading N assemblies will consume RAM. But why it consumes 10-folds for the same list, but read in parallel?

**Comment by jbevain** [comment]
It's hard to tell, by doing things in parallel you're adding pressure to the GC to allocate, it's possible the GC is delaying collections and using the full extent of available memory to satisfy that pressure.

**Comment by SimonCropp** [comment]
@nefcanto 

firstly you have a bug. List is not thread safe. use ConcurrentBag instead

Secondly i suspect you may be wasting ram by not caching resolved assembliesassembly. try this


    static List<Tuple<AssemblyDefinition, string>> ConvertToReflectiveAssemblies(List<string> assemblyFiles)
    {
        var temp = new ConcurrentBag<Tuple<AssemblyDefinition, string>>();

        var readerParameters = new ReaderParameters(ReadingMode.Deferred) {AssemblyResolver = new ThreadSafeAssemblyResolver()};

        Parallel.ForEach(assemblyFiles, currentFile =>
        {
            var assembly = AssemblyDefinition.ReadAssembly(currentFile, readerParameters);
            temp.Add(new Tuple<AssemblyDefinition, string>(assembly, currentFile));
        });

        return temp.ToList();
    }

    public class ThreadSafeAssemblyResolver : BaseAssemblyResolver
    {
        ConcurrentDictionary<string, AssemblyDefinition> cache;

        public ThreadSafeAssemblyResolver()
        {
            cache = new ConcurrentDictionary<string, AssemblyDefinition>(StringComparer.Ordinal);
        }

        public override AssemblyDefinition Resolve(AssemblyNameReference name)
        {
            return cache.GetOrAdd(name.FullName, s => base.Resolve(name));
        }

        protected override void Dispose(bool disposing)
        {
            foreach (var value in cache.Values)
            {
                value.Dispose();
            }

            cache.Clear();
            base.Dispose(disposing);
        }
    }
